File: edit_with_undo.py
from pathlib import Path
import shutil

def edit_with_undo(file_path, edit_func, config, *args, **kwargs):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    undo_path = file_path.with_suffix(file_path.suffix + ".undo")
    shutil.copy(file_path, undo_path)
    edit_func(file_path, config, *args, **kwargs)
    return f"Edit applied to {file_path}, undo available at {undo_path}"

def undo_edit(file_path, config):
    undo_path = file_path.with_suffix(file_path.suffix + ".undo")
    if undo_path.exists():
        shutil.move(undo_path, file_path)
        return f"Restored {file_path} from undo"
    return "No undo available"

File: ts-tools/edit_with_undo.ts
import * as fs from 'fs';
import * as path from 'path';
import { checkFilePermission } from './check_file_permission';

export function editWithUndo(
  filePath: string,
  editFunc: (filePath: string, config: any, ...args: any[]) => Promise<string>,
  config: any,
  ...args: any[]
): Promise<string> {
  return new Promise((resolve, reject) => {
    checkFilePermission(filePath, 'write', config)
      .then(() => {
        const resolvedPath = path.resolve(filePath);
        const undoPath = resolvedPath + '.undo';
        fs.copyFile(resolvedPath, undoPath, (err) => {
          if (err) {
            reject(err);
          } else {
            editFunc(filePath, config, ...args)
              .then(() => {
                resolve(
                  `Edit applied to ${filePath}, undo available at ${undoPath}`,
                );
              })
              .catch(reject);
          }
        });
      })
      .catch(reject);
  });
}

export function undoEdit(filePath: string, config: any): Promise<string> {
  return new Promise((resolve, reject) => {
    const resolvedPath = path.resolve(filePath);
    const undoPath = resolvedPath + '.undo';
    fs.access(undoPath, fs.constants.F_OK, (err) => {
      if (err) {
        resolve('No undo available');
      } else {
        fs.rename(undoPath, resolvedPath, (err) => {
          if (err) {
            reject(err);
          } else {
            resolve(`Restored ${filePath} from undo`);
          }
        });
      }
    });
  });
}
